### PR TITLE
GEODE-6470: Making double-checked locking thread-safe

### PR DESCRIPTION
**Pull request by nabarunnag**
* A repeated check on a non-volatile field is not thread-safe, and could result in unexpected behavior.
	* Fixed using volatile and local variables

Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [ ] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [ ] Is your initial contribution a single, squashed commit?

- [ ] Does `gradlew build` run cleanly?

- [ ] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

### CODE REVIEW

File: geode-core/src/main/java/org/apache/geode/internal/cache/wan/GatewaySenderEventImpl.java [hunk 1]
```diff
@@ -117,7 +117,7 @@
   /**
    * The serialized new value for this event's key. May not be computed at construction time.
    */
-  protected byte[] value;
+  protected volatile byte[] value;
```
Comment (jdeppe-pivotal) [inline geode-core/src/main/java/org/apache/geode/internal/cache/wan/GatewaySenderEventImpl.java]: Did you mean to make this `volatile`? It seems unrelated.

File: geode-core/src/main/java/org/apache/geode/internal/datasource/GemFireBasicDataSource.java [hunk 2]
```diff
@@ -36,7 +36,7 @@
   private static final long serialVersionUID = -4010116024816908360L;
 
   /** Creates a new instance of BaseDataSource */
-  protected transient Driver driverObject = null;
+  protected transient volatile Driver driverObject = null;
```
Comment (jdeppe-pivotal) [inline geode-core/src/main/java/org/apache/geode/internal/datasource/GemFireBasicDataSource.java]: Did you mean to make this `volatile`? It seems unrelated.

File: geode-core/src/main/java/org/apache/geode/internal/cache/wan/GatewaySenderEventImpl.java (hunk 1 above)
Comment (nabarunnag) [reply geode-core/src/main/java/org/apache/geode/internal/cache/wan/GatewaySenderEventImpl.java]: The double checking is done in getSerializedValue, the local variable use was present but the value was not volatile, which is recommended post jdk5

File: geode-core/src/main/java/org/apache/geode/internal/datasource/GemFireBasicDataSource.java (hunk 2 above)
Comment (nabarunnag) [reply geode-core/src/main/java/org/apache/geode/internal/datasource/GemFireBasicDataSource.java]: @jdeppe-pivotal 
The double checking is done in getConnection, Updating to use local variables too
